cwl/src/main/scala/wdl4s/cwl/WorkflowStep.scala

```diff
@@ -1,45 +1,189 @@
+      case (id, tpe) => RunOutputId(id).outputId -> tpe
+    }
+
+    out.fold(WorkflowOutputsToOutputDefinition).apply(idMap)
+  }
+
+  def taskDefinition(typeMap: WdlTypeMap, cwlMap: Map[String, CwlFile]): TaskDefinition = {
+
+    val id = this.id
+
+    val commandTemplate: Seq[CommandPart] =
+    //TODO: turn this select into a fold that supports other types of runnables
+      run.select[CommandLineTool].map {
+        clt =>
+          clt.baseCommand.map(_.fold(BaseCommandToCommandParts)).toSeq.flatten ++
+            clt.arguments.map(_.map(_.fold(ArgumentToCommandPart)).toSeq).toSeq.flatten
+      }.toSeq.flatten
+
+    val runtimeAttributes: RuntimeAttributes = RuntimeAttributes(Map.empty[String, WdlExpression])
+
+    val meta: Map[String, String] = Map.empty
+    val parameterMeta: Map[String, String] = Map.empty
+
+    val declarations: List[(String, WomExpression)] = List.empty
+
+    TaskDefinition(
+      id,
+      commandTemplate,
+      runtimeAttributes,
+      meta,
+      parameterMeta,
+      taskDefinitionOutputs(cwlMap),
+      taskDefinitionInputs(typeMap),
+      declarations
+    )
+  }
+
+  /**
+    * In order to produce a map to lookup workflow
+    * @param typeMap
+    * @param cwlMap
+    * @param workflow
+    * @return
+    */
+  def callWithInputs(typeMap: WdlTypeMap,
+                     cwlMap: Map[String, CwlFile],
+                     workflow: Workflow,
+                     knownNodes: Set[GraphNode],
+                     workflowInputs: Map[String,
+                       GraphNodeOutputPort]): Set[GraphNode] = {
+
+    //TODO: Check to see if this step is already present in the graphNodes
+
+    val haveWeSeenThisStep = knownNodes.flatMap{_ match {
+      case TaskCallNode(name, _, _) => Set(name)
+      case _ => Set.empty[String]
+    }}.contains(id)
+
+    if (haveWeSeenThisStep)
+      knownNodes
+    else {
+      //need the outputs mapped from other workflow steps in order to pass in this map
+      val workflowOutputsMap: (Map[String, OutputPort], Set[GraphNode]) = in.foldLeft((Map.empty[String, OutputPort], knownNodes)) {
+        case ((map, upstreamNodes), workflowStepInput) =>
+
+          val inputSource = workflowStepInput.source.flatMap(_.select[String]).get
+
+          FullyQualifiedName(inputSource) match {
+            case WorkflowStepOutputIdReference(_, _, stepId) =>
+
+              //generate a node for this output
+
+              def findThisInputInSet(set: Set[GraphNode]) ={
+
+                val lookupSet: Set[OutputPort] =
+                  for {
+                    node <- set
+                    outputPort <- node.outputPorts
+                  } yield outputPort
+
+                lookupSet.find(_.name == inputSource)
+              }
+
+              val foundInSeenNodes: Option[OutputPort] = findThisInputInSet(knownNodes)
+
+              def lookupUpstreamNodes:Option[(Set[GraphNode], OutputPort)] = {
+                //TODO: Option get!
```

Comment:
I guess you could throw one in if you really like but that seems kind of strange